### PR TITLE
[20.10 backport] [deb] fixes for version generating

### PR DESCRIPTION
**Pull request by thaJeztah**
Backports:

- https://github.com/docker/docker-ce-packaging/pull/677
- https://github.com/docker/docker-ce-packaging/pull/700
- https://github.com/docker/docker-ce-packaging/pull/701
    - :warning: not including the last commit: https://github.com/docker/docker-ce-packaging//commit/f8299f20216a4e2def617f7d3605eab1d68132ee

### CODE REVIEW

File: deb/gen-deb-ver [hunk 1]
```diff
@@ -13,39 +13,20 @@ GIT_COMMAND="git -C $REPO_DIR"
...
-		debVersion="$(gen_deb_version "$debVersion" tp 0)"
-		;;
-	*-beta[.0-9]*)
-		debVersion="$(gen_deb_version "$debVersion" beta 1)"
-		;;
-	*-rc[.0-9]*)
-		debVersion="$(gen_deb_version "$debVersion" rc 2)"
-		;;
-	*)
-		debVersion="$debVersion-3"
```
Comment (thaJeztah) [inline deb/gen-deb-ver:43]: Need to make some changes to preserve this `~3` for the 20.10 branch (for now) so that the `docker-ce`, `docker-ce-cli`, and `docker-ce-rootless-extras` packages keep it (23.0.0 will start using the new versioning scheme)

File: deb/build-deb [hunk 2]
```diff
@@ -40,8 +40,11 @@ debSource="$(awk -F ': ' '$1 == "Source" { print $2; exit }' debian/control)"
 debMaintainer="$(awk -F ': ' '$1 == "Maintainer" { print $2; exit }' debian/control)"
 debDate="$(date --rfc-2822)"
 
+# Note: on the 20.10 branch, we preserve the "~3" suffix that was removed in
+#       commit 8a720b28a5d0db8fc93423cca79ca21b9b69f1f1, and continue using the
+#       old version scheme for docker-ce, docker-ce-cli, and docker-rootless-extras.
 cat > "debian/changelog" <<-EOF
-$debSource (${EPOCH}${EPOCH_SEP}${DEB_VERSION}-0~${DISTRO}-${SUITE}) $SUITE; urgency=low
+$debSource (${EPOCH}${EPOCH_SEP}${DEB_VERSION}~3-0~${DISTRO}-${SUITE}) $SUITE; urgency=low
```
Comment (thaJeztah) [inline deb/build-deb:47]: Revert the changes from 8a720b28a5d0db8fc93423cca79ca21b9b69f1f1 for the docker-ce packages (but continue using the new format for the plugins, so that we don't push the same version of a plugin twice (once through the 23.0 release, once through the 20.10 release).

File: rpm/gen-rpm-ver [hunk 3]
```diff
@@ -42,7 +42,12 @@ rpmVersion="${VERSION#v}"
 # Docker 22.06.0-beta.1:  version=22.06.0, release=1
 # Docker 22.06.0-rc.1:    version=22.06.0, release=1
 # Docker 22.06.0-dev:     version=0.0.0~YYYYMMDDHHMMSS.gitHASH, release=0
-rpmRelease=1
+# rpmRelease=1
+
+# Note: on the 20.10 branch, continue using "3" as default, which was removed in
+#       commit 8a720b28a5d0db8fc93423cca79ca21b9b69f1f1. 23.0.0 and up use "1"
+#       as default.
+rpmRelease=3
```
Comment (thaJeztah) [inline rpm/gen-rpm-ver:50]: Same here, for RPMs

**Review by vvoland — APPROVED** [review]

**Comment by thaJeztah** [comment]
Thx! Let me bring this one in

File: deb/build-deb (hunk 2 above)
Comment (tianon) [reply deb/build-deb:47]: Wait, what's the `~3` here for?  (Sorry, I'm missing some context :see_no_evil:)

Comment (thaJeztah) [reply deb/build-deb:47]: This was used in 20.10 (and before) to.. make sure "stable" releases ranked higher than "pre-releases" (`:lolsob:`). I removed that code from master, but the same generation code is used for all deb versions, so I backported removing that ugly hack in this PR (see https://github.com/docker/docker-ce-packaging/pull/831/commits/8a720b28a5d0db8fc93423cca79ca21b9b69f1f1). However, as 20.10 is reaching its end, I didn't want to introduce changing the versioning scheme for it, so I added it back here. As we're not planning to do `-beta` releases for 20.10 patch releases, I hardcoded it to `~3` (which was used to designate "stable releases').
